### PR TITLE
Styling refinements to people box and spectrum

### PR DESCRIPTION
**Pull request by aldeka**
* Turned the spectrum bootstrap blue ("primary") for consistency
* Turned the icons bootstrap red ("danger")
* Moved icons  a bit further away from the initial for readability
* Added additional style changes for boxes that are deafened

### CODE REVIEW

**Review by grantaelliott — COMMENTED** [review]
This looks good. I left some suggestions below - mostly pertaining to style. (Feel free to disagree.)

File: imports/client/components/CallLinkBox.tsx [hunk 1]
```diff
@@ -274,22 +274,24 @@ class CallLinkBox extends React.PureComponent<CallLinkBoxProps, CallLinkBoxState
...
           key={`viewer-${this.props.peerParticipant._id}`}
-          className="people-item"
+          className={`people-item ${this.props.peerParticipant.muted ? 'muted' : ''} ${this.props.peerParticipant.deafened ? 'deafened' : ''} ${this.props.peerParticipant.muted || this.props.peerParticipant.deafened ? '' : 'active'}`}
         >
           <span className="initial">{name.slice(0, 1)}</span>
           <div className="webrtc">
+            {this.props.peerParticipant.muted && <span className="icon muted-icon"><FontAwesomeIcon icon={faMicrophone} /></span>}
+            {this.props.peerParticipant.deafened && <span className="icon deafened-icon"><FontAwesomeIcon icon={faHeadphonesAlt} /></span>}
+            {!this.props.spectraDisabled &&
+              !this.props.peerParticipant.muted &&
```
Comment (zarvox) [inline imports/client/components/CallLinkBox.tsx:284]: This doesn't reliably come back when the peer mutes and unmutes, because `Spectrum` needs to have `connect` called on it to provide it with a stream.  (To the extent that it does work by coincidence, it must be relying on object reuse implementation details that are far from guaranteed.)

The other Spectrum instance is fine because it uses a callback ref to ensure we propagate the stream that we already have laying around at the time of rendering; it's possible we could do so here if we have the track already but we'd still need to save a handle since we might not have the stream/track from the RTCPeerConnection yet.  Similarly, we'd need to modify the onRemoteTrack callback to do the same "check if we have both objects, connect if so" logic.

**Review by zarvox — CHANGES_REQUESTED** [review]

File: imports/client/components/CallLinkBox.tsx (hunk 1 above)
Comment (aldeka) [reply imports/client/components/CallLinkBox.tsx:284]: I don't really understand this comment? The Spectrum component still has the ref passed to it, it's just further down the file. And the icons already refer to `peerParticipant.muted`...

File: imports/client/components/CallLinkBox.tsx [hunk 2]
```diff
@@ -274,22 +274,24 @@ class CallLinkBox extends React.PureComponent<CallLinkBoxProps, CallLinkBoxState
...
+              !this.props.peerParticipant.deafened ? (
```
[the remark this answers is not in the document]
Comment (aldeka) [reply imports/client/components/CallLinkBox.tsx:285]: I wasn't certain if we could rely on `muted` to always be true in the deafened case?

File: client/stylesheets/puzzle.scss [hunk 3]
```diff
@@ -367,8 +367,53 @@ table.puzzle-list {
...
+
     .initial {
-      z-index: 10
+      z-index: 10;
+    }
+
+    &.active .initial {
+      // white border around initial
+      // aids readability if the spectrum graph is high
+      text-shadow: -1px 1px 0 white,
```
[the remark this answers is not in the document]
Comment (aldeka) [reply client/stylesheets/puzzle.scss:413]: I meant the border to only appear in the active case because the initial was a little hard to read over the blue of the spectrum. IMO the initial has plenty of contrast in the deafened case, but if you disagree I can add it across the board.

File: client/stylesheets/puzzle.scss [hunk 4]
```diff
@@ -367,8 +367,53 @@ table.puzzle-list {
...
+
+    &.muted .muted-icon {
+      position: absolute;
+      top: $mutedTop;
+      right: $mutedRight;
+    }
+
+    &.deafened {
+      color: $secondary;
+      background: #ffffff88;
```
[the remark this answers is not in the document]
Comment (aldeka) [reply client/stylesheets/puzzle.scss:390]: It's not, exactly? It's just translucent white, it'll show whatever the section background happens to be faintly.

File: client/stylesheets/puzzle.scss [hunk 5]
```diff
@@ -367,8 +367,53 @@ table.puzzle-list {
...
+      font-size: 12px;
+      width: 16px;
+      text-align: center;
+      color: $danger;
+      height: 50%;
+    }
+
+    $mutedTop: -4px;
+    $mutedRight: -2px;
+    $deafenDiff: 2px;
```
[the remark this answers is not in the document]
Comment (aldeka) [reply client/stylesheets/puzzle.scss:380]: It would be simpler to have the border all the time, except I'd have to frob the spectrum width for it to still go edge to edge -- it looks kind of funky otherwise.

File: client/stylesheets/puzzle.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (grantaelliott) [reply client/stylesheets/puzzle.scss:390]: Yeah, I was just curious about the choice. I suppose it depicts that they’re only sort-of present, which makes sense.

File: client/stylesheets/puzzle.scss (hunk 5 above)
[the remark this answers is not in the document]
Comment (grantaelliott) [reply client/stylesheets/puzzle.scss:380]: Yeah, my preference would tend towards that. It still has an awkward offset but eliminates the state dependence.

File: imports/client/components/CallLinkBox.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (grantaelliott) [reply imports/client/components/CallLinkBox.tsx:285]: I’m pretty sure the decoding has already happened by this point (after all, the muted icon just above depends solely on the muted prop), so in practice the deafened term is redundant. But if the two were to be separated again, my expectation is that only the muting would matter for hiding the spectrum.

File: client/stylesheets/puzzle.scss (hunk 3 above)
[the remark this answers is not in the document]
Comment (grantaelliott) [reply client/stylesheets/puzzle.scss:413]: Sorry, I think I forgot the intent of this someplace between my first and second sentence... I’ll try again:

If spectrum hiding is just done based on the muted flag (see below), then this can just be `:not(.muted)` and we can get rid of the active class entirely.

File: imports/client/components/CallLinkBox.tsx (hunk 1 above)
Comment (zarvox) [reply imports/client/components/CallLinkBox.tsx:284]: Sorry, let me try again.

Just putting a `<Spectrum>` in the component tree will not get you any visible spectra, because there's no media stream being fed into it yet.  We feed in that media stream in exactly one place in this file; in the `onNewRemoteTrack` callback.  So if the `Spectrum` element is not around when that single callback fires, or vanishes and returns, then the Spectrum element will have no audio feed to visualize.

Repro steps:

1. Have two callers join a call, call them Participant A and Participant B.
2. Participant B mutes or deafens.
3. Participant A sees it and removes the `<Spectrum>` from the render.
4. Participant B unmutes/undeafens.
5. Participant A sees it and adds `<Spectrum>` to the render again
6. But nothing ever calls `.connect` on the instance of Spectrum in step 5.
7. So Participant A sees only silence in the spectrum drawn for Participant B forevermore.

The reason this isn't *also* an issue in CallSection is because we always have the local microphone stream in advance of the Spectrum getting rendered, so we call `.connect` in the callback ref in CallSection at https://github.com/deathandmayhem/jolly-roger/pull/343/files#diff-a9128bbe54b44b926e42180c5397ea0a33dd7256546bf5fd524b355c7fde60b6R90

There's a couple options for fixing this:

1. Make Spectrum take a stream as a prop from its parent, and carefully wire up the analyserNode in componentDidMount/componentDidUpdate.  We'd need to hold on to the `stubStream` in `CallLinkBox`.
2. Alternately, we could make `CallLinkBox` track when both the `Spectrum` and the stream are ready, and call `spectrumRef.current.connect(theStream)` in each case where appropriate.  That's what I tried to propose in the previous comment.

Comment (zarvox) [reply imports/client/components/CallLinkBox.tsx:284]: To add additional comment: this bug is fundamentally my fault and was always an issue if we toggled the `spectraDisabled` feature flag too!  I just figured that probably wouldn't come up in practice with the feature flag, since we probably don't toggle that a whole lot (and if we turn it off we probably leave it off).  Having this break on mute/unmute seems like substantially more exposure.

Worst case I can always take this patch and just take a crack at fixing the bug myself

**Comment by zarvox** [comment]
Wound up cherry-picking this as https://github.com/deathandmayhem/jolly-roger/pull/343/commits/555c2b4c93c9075d49ce6e0496dc502a9859d915 and following up on some of the comments in subsequent commits.

I think there's a couple remaining suggestions from Grant.  I'm gonna close this PR, but feel free to reopen or file a new one.
